sql/sort.go

```diff
@@ -0,0 +1,247 @@
+			target := string(t.Base)
+			for j, col := range columns {
+				if equalName(target, col) {
+					index = j + 1
+					break
+				}
+			}
+			if index == 0 {
+				return nil, fmt.Errorf("qualified name \"%s\" not found", t)
+			}
+
+		default:
+			// TODO(pmattis): Use NormalizeExpr.
+			datum, err := parser.EvalExpr(o.Expr)
+			if err != nil {
+				if err := s.addRender(parser.SelectExpr{Expr: o.Expr}); err != nil {
+					return nil, err
+				}
+				index = len(s.columns)
+				break
+			}
+			i, ok := datum.(parser.DInt)
+			if !ok {
+				return nil, fmt.Errorf("invalid order by (%s)", t)
+			}
+			index = int(i)
+			if index < 1 || index > len(columns) {
+				return nil, fmt.Errorf("invalid order by index: %d not in range [1, %d]",
+					index, len(columns))
+			}
+		}
+
+		if o.Direction == parser.Descending {
+			index = -index
+		}
+		ordering = append(ordering, index)
+	}
+
+	return &sortNode{columns: columns, ordering: ordering}, nil
+}
+
+type sortNode struct {
+	plan     planNode
+	columns  []string
+	ordering []int
+	values   *valuesNode
+	err      error
+}
+
+func (n *sortNode) Columns() []string {
+	return n.columns
+}
+
+func (n *sortNode) Ordering() []int {
+	return n.ordering
+}
+
+func (n *sortNode) Values() parser.DTuple {
+	// If an ordering expression was used the number of columns in each row might
+	// differ from the number of columns requested, so trim the result.
+	v := n.values.Values()
+	return v[:len(n.columns)]
+}
+
+func (n *sortNode) Next() bool {
+	if n.values == nil {
+		if !n.initValues() {
+			return false
+		}
+	}
+	return n.values.Next()
+}
+
+func (n *sortNode) Err() error {
+	return n.err
+}
+
+// wrap the supplied planNode with the sortNode if sorting is required.
+func (n *sortNode) wrap(plan planNode) planNode {
+	if n == nil {
+		return plan
+	}
+
+	// Check to see if the requested ordering is compatible with the existing
+	// ordering.
+	existingOrdering := plan.Ordering()
+	for i := range n.ordering {
+		if i >= len(existingOrdering) || n.ordering[i] != existingOrdering[i] {
+			n.plan = plan
+			return n
+		}
+	}
+	return plan
+}
+
+func (n *sortNode) initValues() bool {
+	// TODO(pmattis): If the result set is large, we might need to perform the
+	// sort on disk.
+	n.values = &valuesNode{}
+	for n.plan.Next() {
+		values := n.plan.Values()
+		valuesCopy := make(parser.DTuple, len(values))
+		copy(valuesCopy, values)
+		n.values.rows = append(n.values.rows, valuesCopy)
+	}
+	n.err = n.plan.Err()
+	if n.err != nil {
+		return false
+	}
+	sort.Sort(n)
+	return true
+}
+
+func (n *sortNode) Len() int {
+	return len(n.values.rows)
+}
+
+func (n *sortNode) Less(i, j int) bool {
+	// TODO(pmattis): An alternative to this type of field-based comparison would
+	// be to construct a sort-key per row using encodeTableKey(). Using a
+	// sort-key approach would likely fit better with a disk-based sort.
+	ra, rb := n.values.rows[i], n.values.rows[j]
+	for _, k := range n.ordering {
+		var da, db parser.Datum
```

Comment:
Ok. I've added `Datum.Compare` and provided implementations. This did dramatically simplify this method, pushing the guts into the datum implementations.